local-path-cleaner/blog_post.md

```diff
@@ -0,0 +1,619 @@
+  resources:
+    requests:
+      ephemeral-storage: "5Gi"
+  volumeMounts:
+  - mountPath: /data
+    name: local-storage
+```
+
+The scheduler will take storage requirements into account when scheduling pods. We can use this to avoid overcommitting local storage on a node. However, local ephemeral storage and persistent volumes serve different purposes (one is ephemeral, the other persistent). Kubernetes does not track PVC volumes as ephemeral storage consumption so we cannot combine local ephemeral storage requests and `local-path` PVCs out of the box.
+
+Luckily, with a trick during node provisioning, we can still achieve what we are looking for. Let's consider GCP as an example. In our startup script, we might manually mount multiple local NVMe SSD in a RAID0 device:
+
+```bash
+# Find all SSDs
+SSDs=($(readlink -f /dev/disk/by-id/google-local-nvme-ssd-*))
+
+# Create RAID0 device
+mdadm --create /dev/md0 \
+  --level=0 --force \
+  "--raid-devices=$${#SSDs[@]}" \
+  "$${SSDs[@]}"
+
+# Format RAID0 device
+mkfs.xfs -s size=4096 /dev/md0
+
+# Mount RAID0 device to /mnt/disks/ssd-array
+mkdir -p /mnt/disks/ssd-array
+mount /dev/md0 /mnt/disks/ssd-array
+chmod a+w /mnt/disks/ssd-array
+
+# Create fstab entry (to survive reboots)
+raid_dev_uuid=$(blkid | grep dev/md0 | egrep -o '[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}')
+echo "UUID=$raid_dev_uuid /mnt/disks/ssd-array xfs defaults,nofail,noatime 0 0" |\
+  tee -a /etc/fstab
+  
+# Disable NODE_LOCAL_SSDS_EPHEMERAL as we manage ephemeral storage ourselves
+sed -i 's|readonly NODE_LOCAL_SSDS_EPHEMERAL=true|readonly NODE_LOCAL_SSDS_EPHEMERAL=false|' \
+  "$${KUBE_HOME}/kube-env"
+```
+
+Kubelet tracks ephemeral storage in certain locations. By bind mounting these into our RAID0 mount, we effectively enable Kubernetes to track the capacity of our custom local storage.
+
+```bash
+mkdir -p /mnt/disks/ssd-array/lib/kubelet
+mv /var/lib/kubelet/* /mnt/disks/ssd-array/lib/kubelet
+mount --bind /mnt/disks/ssd-array/lib/kubelet /var/lib/kubelet
+
+mkdir -p /mnt/disks/ssd-array/lib/containerd
+mv /var/lib/containerd/* /mnt/disks/ssd-array/lib/containerd
+mount --bind /mnt/disks/ssd-array/lib/containerd /var/lib/containerd
+
+mkdir -p /mnt/disks/ssd-array/stateful_partition
+mount --bind /mnt/disks/ssd-array/stateful_partition /mnt/stateful_partition
+```
+
+Alternatively, we could hard code the available ephemeral storage capacity in the kubelet config based on the available space on the RAID0 device. This would allow the scheduler to take storage requests into account for your `local-path` PVs. If you wanted it to be 500Gi, you could run:
+
+```bash
+sed -i -E 's/(ephemeral-storage:).*/\1 500Gi/' /home/kubernetes/kubelet-config.yaml
+```
+
+When querying the node capacity, you should see the ephemeral storage capacity reflected:
+
+```yaml
+status:
+  capacity:
+    cpu: "16"
+    ephemeral-storage: 500Gi
+    memory: 128Gi
+    pods: "110"
+```
```

Comment:
Does this work if we were to add multiple PVCs to the node, each w/ < 500Gi requested? i.e. can k8s track how much of this capacity was committed? 

My guess is this is checked statically when scheduling: the requested PVC capacity must be less than the node capacity. But if I'd already bound 5 100gb PVCs to the node, the 6th would still pass this check.

Reply:
I've misunderstood the options presented. The goal is to inform k8s how much disk capacity is being used, by:
1) bind mounting the default locations the kubelet checks; or
2) explicitly setting the capacity in the kubelet config.

In both options, we still need to declare the `ephemeralStorage`. This is how the tracking is done.